File: Season-2/Level-4/code.js
// Welcome to Secure Code Game Season-2/Level-4!

// Follow the instructions below to get started:

// 1. test.js is passing but the code here is vulnerable
// 2. Review the code. Can you spot the bugs(s)?
// 3. Fix the code.js but ensure that test.js passes
// 4. Run hack.js and if passing then CONGRATS!
// 5. If stuck then read the hint
// 6. Compare your solution with solution.js

const express = require("express");
const bodyParser = require("body-parser");
const libxmljs = require("libxmljs");
const multer = require("multer");
const app = express();

app.use(bodyParser.json());
app.use(bodyParser.text({ type: "application/xml" }));

const storage = multer.memoryStorage();
const upload = multer({ storage });

app.post("/ufo/upload", upload.single("file"), (req, res) => {
  return res.status(501).send("Not Implemented.");
  // We don't need this feature/endpoint, it's a backdoor! 
  // Removing this prevents an attacker to perform a Remote Code Execution
  // by uploading a file with a .admin extension that is then executed on the server.
  // The best code is less code. If you don't need something, don't include it.
});

app.post("/ufo", (req, res) => {
  const contentType = req.headers["content-type"];

  if (contentType === "application/json") {
    console.log("Received JSON data:", req.body);
    res.status(200).json({ ufo: "Received JSON data from an unknown planet." });
  } else if (contentType === "application/xml") {
    try {
      const xmlDoc = libxmljs.parseXml(req.body, {
        replaceEntities: false, // Disabled the option to replace XML entities
        recover: false, // Disabled the parser to recover from certain parsing errors
        nonet: true, // Disabled network access when parsing
      });

      console.log("Received XML data from XMLon:", xmlDoc.toString());

      const extractedContent = [];

      xmlDoc
        .root()
        .childNodes()
        .forEach((node) => {
          if (node.type() === "element") {
            extractedContent.push(node.text());
          }
        });

      if (
        xmlDoc.toString().includes('SYSTEM "') &&
        xmlDoc.toString().includes(".admin")
      ) {
        // Removed the code to execute commands within the .admin file on the server
        res.status(400).send("Invalid XML");         
      } else {
        res
          .status(200)
          .set("Content-Type", "text/plain")
          .send(extractedContent.join(" "));
      }
    } catch (error) {
      console.error("XML parsing or validation error");
      res.status(400).send("Invalid XML");
    }
  } else {
    res.status(405).send("Unsupported content type");
  }
});

const PORT = process.env.PORT || 3000;
const server = app.listen(PORT, () => {
  console.log(`Server is running on port ${PORT}`);
});

module.exports = server;

File: Season-1/Level-1/code.py
'''
Welcome to Secure Code Game Season-1/Level-1!

Follow the instructions below to get started:

1. tests.py is passing but code.py is vulnerable
2. Review the code. Can you spot the bug?
3. Fix the code but ensure that tests.py passes
4. Run hack.py and if passing then CONGRATS!
5. If stuck then read the hint
6. Compare your solution with solution.py
'''

from collections import namedtuple

Order = namedtuple('Order', 'id, items')
Item = namedtuple('Item', 'type, description, amount, quantity')

MAX_ORDER_AMOUNT = 10000
MAX_PRODUCT_QUANTITY = 1000
tolerance = 0.01

def validorder(order: Order):
    net = 0
    product_quantity = 0
    product_amount = 0
    for item in order.items:
        if item.type == 'payment':
            net += item.amount
        elif item.type == 'product':
            # net -= item.amount * item.quantity
            product_quantity += item.quantity
            product_amount += item.amount * item.quantity
        else:
            return "Invalid item type: %s" % item.type
    
    if product_amount > MAX_ORDER_AMOUNT:
        return "Total amount payable for an order exceeded"
    
    if abs(net - product_amount) > tolerance:
        return "Order ID: %s - Payment imbalance: $%0.2f" % (order.id, net - product_amount)
    
    if product_quantity > MAX_PRODUCT_QUANTITY:
        return "Total quantity of products exceeded"

    return "Order ID: %s - Full payment received!" % order.id